### ISSUE TITLE
prov/psm3: covscan report from redhat

### ISSUE THREAD

**Issue by shefty** [ofiwg/libfabric#6819]
```
Error: CHECKED_RETURN (CWE-252): [#def77]
libfabric-1.12.1/prov/psm2/src/psmx2_trx_ctxt.c:111: check_return: Calling "pthread_create" without checking return value (as is done elsewhere 15 out of 18 times).
libfabric-1.12.1/prov/psm2/src/psmx2_domain.c:141: example_assign: Example 1: Assigning: "err" = return value from "pthread_create(&domain->progress_thread, NULL, psmx2_progress_func, (void *)domain)".
libfabric-1.12.1/prov/psm2/src/psmx2_domain.c:143: example_checked: Example 1 (cont.): "err" has its value checked in "err".
libfabric-1.12.1/prov/psm2/src/psmx2_wait.c:118: example_assign: Example 2: Assigning: "err" = return value from "pthread_create(&psmx2_wait_thread, &attr, psmx2_wait_progress, (void *)fabric)".
libfabric-1.12.1/prov/psm2/src/psmx2_wait.c:120: example_checked: Example 2 (cont.): "err" has its value checked in "err".
libfabric-1.12.1/prov/psm3/psm3/psm_mq.c:1413: example_checked: Example 3: "pthread_create(&mq->mq_perf_data.perf_print_thread, NULL, psmi_mq_print_stats_thread, (void *)mq)" has its value checked in "pthread_create(&mq->mq_perf_data.perf_print_thread, NULL, psmi_mq_print_stats_thread, (void *)mq)".
libfabric-1.12.1/prov/psm3/psm3/psm_stats.c:289: example_checked: Example 4: "pthread_create(&perf_print_thread, NULL, psmi_print_stats_thread, NULL)" has its value checked in "pthread_create(&perf_print_thread, NULL, psmi_print_stats_thread, NULL)".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ptl_rcvthread.c:145: example_checked: Example 5: "pthread_create(&rcvc->hdrq_threadid, NULL, ips_ptl_pollintr, ptl->rcvthread)" has its value checked in "pthread_create(&rcvc->hdrq_threadid, NULL, ips_ptl_pollintr, ptl->rcvthread)".
#  109|   			disconn->trx_ctxt = trx_ctxt;
#  110|   			disconn->epaddr = epaddr;
#  111|-> 			pthread_create(&disconnect_thread, NULL,
#  112|   				       disconnect_func, disconn);
#  113|   			pthread_detach(disconnect_thread);

Error: TOCTOU (CWE-367): [#def78]
libfabric-1.12.1/prov/psm3/psm3/libuuid/psm_uuid.c:83: fs_check_call: Calling function "stat" to perform check on ""/dev/urandom"".
libfabric-1.12.1/prov/psm3/psm3/libuuid/psm_uuid.c:88: toctou: Calling function "open" that uses ""/dev/urandom"" after a check function. This can cause a time-of-check, time-of-use race condition.
#   81|   	size_t nbytes;
#   82|   	int fd;
#   83|-> 	if(stat("/dev/urandom", &urandom_stat) != 0) {
#   84|   		psmi_make_drand_uuid(uuid_out);
#   85|   		return;

Error: CLANG_WARNING: [#def79]
libfabric-1.12.1/prov/psm3/psm3/opa/opa_syslog.c:96:2: warning[deadcode.DeadStores]: Value stored to 'len' is never read
#   94|   	}
#   95|   
#   96|-> 	len = snprintf(logprefix, sizeof(logprefix),
#   97|   		       "(nic/%s)[%d]: %s", prefix ? prefix : "nic",
#   98|   		       (int)getpid(), format);

Error: NULL_RETURNS (CWE-476): [#def80]
libfabric-1.12.1/prov/psm3/psm3/psm.c:823: returned_null: "sysfs_unit_path" returns "NULL" (checked 6 out of 7 times).
libfabric-1.12.1/prov/psm3/psm3/psm.c:823: dereference: Dereferencing a pointer that might be "NULL" "sysfs_unit_path(unit)" when calling "strrchr".
libfabric-1.12.1/prov/psm3/psm3/opa/opa_sysfs.c:248: example_assign: Example 1: Assigning: "unitpath" = return value from "sysfs_unit_path(unit)".
libfabric-1.12.1/prov/psm3/psm3/opa/opa_sysfs.c:250: example_checked: Example 1 (cont.): "unitpath" has its value checked in "unitpath == NULL".
libfabric-1.12.1/prov/psm3/psm3/opa/opa_sysfs.c:193: example_assign: Example 2: Assigning: "unitpath" = return value from "sysfs_unit_path(unit)".
libfabric-1.12.1/prov/psm3/psm3/opa/opa_sysfs.c:195: example_checked: Example 2 (cont.): "unitpath" has its value checked in "unitpath == NULL".
libfabric-1.12.1/prov/psm3/psm3/psm.c:818: example_assign: Example 3: Assigning: "pathName" = return value from "sysfs_unit_path(unit)".
libfabric-1.12.1/prov/psm3/psm3/psm.c:821: example_checked: Example 3 (cont.): "pathName" has its value checked in "pathName".
libfabric-1.12.1/prov/psm3/psm3/psm.c:836: example_assign: Example 4: Assigning: "pathName" = return value from "sysfs_unit_path(unit)".
libfabric-1.12.1/prov/psm3/psm3/psm.c:839: example_checked: Example 4 (cont.): "pathName" has its value checked in "pathName".
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_ep.c:1389: example_assign: Example 5: Assigning: "unitpath" = return value from "sysfs_unit_path(unit)".
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_ep.c:1397: example_checked: Example 5 (cont.): "unitpath" has its value checked in "unitpath".
#  821|   			if (!pathName) break;
#  822|   
#  823|-> 			unitName = strrchr(sysfs_unit_path(unit),'/');
#  824|   			if (!unitName) break;
#  825|   

Error: DEADCODE (CWE-561): [#def81]
libfabric-1.12.1/prov/psm3/psm3/psm.c:687: cond_at_least: Condition "q < 0U", taking false branch. Now the value of "q" is at least 0.
libfabric-1.12.1/prov/psm3/psm3/psm.c:687: cond_between: Condition "q >= PSM2_INFO_QUERY_LAST", taking false branch. Now the value of "q" is between 0 and 14.
libfabric-1.12.1/prov/psm3/psm3/psm.c:694: between: When switching on "q", the value of "q" must be between 0 and 14.
libfabric-1.12.1/prov/psm3/psm3/psm.c:694: dead_error_condition: The "switch" governing value "q" cannot reach the "default" case.
libfabric-1.12.1/prov/psm3/psm3/psm.c:846: dead_error_begin: Execution cannot reach this statement: "default:".
#  844|   		}
#  845|   		break;
#  846|-> 	default:
#  847|   		break;
#  848|   	}

Error: CHECKED_RETURN (CWE-252): [#def82]
libfabric-1.12.1/prov/psm3/psm3/psm.c:884: check_return: Calling "psm2_ep_close" without checking return value (as is done elsewhere 4 out of 5 times).
libfabric-1.12.1/prov/psm2/src/psmx2_trx_ctxt.c:219: example_assign: Example 1: Assigning: "err" = return value from "psm2_ep_close(trx_ctxt->psm2_ep, 0, (int64_t)psmx2_env.timeout * 1000000000LL)".
libfabric-1.12.1/prov/psm2/src/psmx2_trx_ctxt.c:224: example_checked: Example 1 (cont.): "err" has its value checked in "err != PSM2_OK".
libfabric-1.12.1/prov/psm2/src/psmx2_trx_ctxt.c:366: example_checked: Example 2: "psm2_ep_close(trx_ctxt->psm2_ep, 0, (int64_t)psmx2_env.timeout * 1000000000LL)" has its value checked in "psm2_ep_close(trx_ctxt->psm2_ep, 0, (int64_t)psmx2_env.timeout * 1000000000LL) != PSM2_OK".
libfabric-1.12.1/prov/psm3/src/psmx3_trx_ctxt.c:219: example_assign: Example 3: Assigning: "err" = return value from "psm2_ep_close(trx_ctxt->psm2_ep, 0, (int64_t)psmx3_env.timeout * 1000000000LL)".
libfabric-1.12.1/prov/psm3/src/psmx3_trx_ctxt.c:224: example_checked: Example 3 (cont.): "err" has its value checked in "err != PSM2_OK".
libfabric-1.12.1/prov/psm3/src/psmx3_trx_ctxt.c:362: example_checked: Example 4: "psm2_ep_close(trx_ctxt->psm2_ep, 0, (int64_t)psmx3_env.timeout * 1000000000LL)" has its value checked in "psm2_ep_close(trx_ctxt->psm2_ep, 0, (int64_t)psmx3_env.timeout * 1000000000LL) != PSM2_OK".
#  882|   	while (ep != NULL) {
#  883|   		psm2_ep_t saved_ep = ep->user_ep_next;
#  884|-> 		psm2_ep_close(ep, PSM2_EP_CLOSE_GRACEFUL,
#  885|   			     2 * PSMI_MIN_EP_CLOSE_TIMEOUT);
#  886|   		psmi_opened_endpoint = ep = saved_ep;

Error: NEGATIVE_RETURNS (CWE-394): [#def83]
libfabric-1.12.1/prov/psm3/psm3/psm.c:918: negative_return_fn: Function "sysconf(_SC_PAGESIZE)" returns a negative number.
libfabric-1.12.1/prov/psm3/psm3/psm.c:918: negative_returns: "sysconf(_SC_PAGESIZE)" is passed to a parameter that cannot be negative.
#  916|   		}
#  917|   
#  918|-> 		msync(shared_affinity_ptr, AFFINITY_SHMEMSIZE, MS_SYNC);
#  919|   
#  920|   		/* End critical section */

Error: NEGATIVE_RETURNS (CWE-394): [#def84]
libfabric-1.12.1/prov/psm3/psm3/psm.c:923: negative_return_fn: Function "sysconf(_SC_PAGESIZE)" returns a negative number.
libfabric-1.12.1/prov/psm3/psm3/psm.c:923: negative_returns: "sysconf(_SC_PAGESIZE)" is passed to a parameter that cannot be negative.
#  921|   		psmi_sem_post(sem_affinity_shm_rw, sem_affinity_shm_rw_name);
#  922|   
#  923|-> 		munmap(shared_affinity_ptr, AFFINITY_SHMEMSIZE);
#  924|   		psmi_free(affinity_shm_name);
#  925|   		affinity_shm_name = NULL;

Error: SIZEOF_MISMATCH (CWE-398): [#def85]
libfabric-1.12.1/prov/psm3/psm3/psm.c:1006: suspicious_sizeof: Passing argument "ctxt" of type "void *" and argument "8UL /* sizeof (void *) */" to function "psm2_setopt" is suspicious.
# 1004|   	/* Eventually deprecate this API to use set/get opt as this is unsafe. */
# 1005|   	PSM2_LOG_MSG("entering");
# 1006|-> 	psm2_setopt(PSM2_COMPONENT_CORE, (const void *)epaddr,
# 1007|   		   PSM2_CORE_OPT_EP_CTXT, (const void *)ctxt, sizeof(void *));
# 1008|   	PSM2_LOG_MSG("leaving");

Error: NEGATIVE_RETURNS (CWE-394): [#def86]
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:158: negative_return_fn: Function "shm_open(affinity_shm_name, 194, 384U)" returns a negative number.
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:158: assign: Assigning: "shm_fd" = "shm_open(affinity_shm_name, 194, 384U)".
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:174: negative_returns: "shm_fd" is passed to a parameter that cannot be negative.
#  172|   	}
#  173|   
#  174|-> 	ret = ftruncate(shm_fd, AFFINITY_SHMEMSIZE);
#  175|   	if ( ret < 0 ) {
#  176|   		_HFI_VDBG("Cannot truncate affinity shared mem fd:%s, errno=%d\n",

Error: NEGATIVE_RETURNS (CWE-394): [#def87]
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:174: negative_return_fn: Function "sysconf(_SC_PAGESIZE)" returns a negative number.
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:174: negative_returns: "sysconf(_SC_PAGESIZE)" is passed to a parameter that cannot be negative.
#  172|   	}
#  173|   
#  174|-> 	ret = ftruncate(shm_fd, AFFINITY_SHMEMSIZE);
#  175|   	if ( ret < 0 ) {
#  176|   		_HFI_VDBG("Cannot truncate affinity shared mem fd:%s, errno=%d\n",

Error: NEGATIVE_RETURNS (CWE-394): [#def88]
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:182: negative_return_fn: Function "sysconf(_SC_PAGESIZE)" returns a negative number.
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:182: negative_returns: "sysconf(_SC_PAGESIZE)" is passed to a parameter that cannot be negative.
#  180|   	}
#  181|   
#  182|-> 	shared_affinity_ptr = (uint64_t *) mmap(NULL, AFFINITY_SHMEMSIZE, PROT_READ | PROT_WRITE,
#  183|   					MAP_SHARED, shm_fd, 0);
#  184|   	if (shared_affinity_ptr == MAP_FAILED) {

Error: NEGATIVE_RETURNS (CWE-394): [#def89]
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:197: negative_return_fn: Function "sysconf(_SC_PAGESIZE)" returns a negative number.
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:197: negative_returns: "sysconf(_SC_PAGESIZE)" is passed to a parameter that cannot be negative. [Note: The source code implementation of the function has been overridden by a builtin model.]
#  195|   		_HFI_VDBG("Creating shm to store NIC affinity per socket\n");
#  196|   
#  197|-> 		memset(shared_affinity_ptr, 0, AFFINITY_SHMEMSIZE);
#  198|   
#  199|   		/*

Error: CLANG_WARNING: [#def90]
libfabric-1.12.1/prov/psm3/psm3/psm_context.c:478:2: warning[deadcode.DeadStores]: Value stored to 'unit_id_prev' is never read
#  476|   	   an expression involving the mod operator. */
#  477|   	int success = 0;
#  478|-> 	unit_id_prev = unit_id = unit_start;
#  479|   	do
#  480|   	{

Error: CLANG_WARNING: [#def91]
libfabric-1.12.1/prov/psm3/psm3/psm_diags.c:175:3: warning[deadcode.DeadStores]: Value stored to 'epaddr' is never read
#  173|   		diags_assert(epaddr->epid == epid_array[i]);
#  174|   		diags_assert(epaddr->ptlctl->ep == ep);
#  175|-> 		epaddr = psmi_epid_remove(ep, epid_array[i]);
#  176|   		epaddr = psmi_epid_lookup(ep, epid_array[i]);
#  177|   		diags_assert(epaddr == NULL);

Error: DEADCODE (CWE-561): [#def92]
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:209: assignment: Assigning: "skip_port" = "1".
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:212: const: At condition "skip_port", the value of "skip_port" must be equal to 1.
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:212: dead_error_condition: The condition "!skip_port" cannot be true.
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:213: dead_error_line: Execution cannot reach this statement: "return psmi_handle_error(NU...".
#  211|   			}
#  212|   			if (! delim && !skip_port)
#  213|-> 				return psmi_handle_error(NULL, PSM2_EP_DEVICE_FAILURE,
#  214|   						"PSM3_MULTIRAIL_MAP invalid format: '%s'",
#  215|   						env_multirail_map.e_str);

Error: UNUSED_VALUE (CWE-563): [#def93]
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:804: value_overwrite: Overwriting previous write to "opts.path_res_type" with value "PSM2_PATH_RES_NONE".
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:785: assigned_value: Assigning value from "opts_i->path_res_type" to "opts.path_res_type" here, but that stored value is overwritten before it can be used.
#  783|   			opts.service_id = (uint64_t) opts_i->service_id;
#  784|   		if (opts_i->path_res_type != PSM2_PATH_RES_NONE)
#  785|-> 			opts.path_res_type = opts_i->path_res_type;
#  786|   
#  787|   		if (opts_i->senddesc_num)

Error: CLANG_WARNING: [#def94]
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:1043:3: warning[deadcode.DeadStores]: Value stored to 'ptl_sizes' is never read
# 1041|   	if (psmi_ep_device_is_enabled(ep, PTL_DEVID_SELF)) {
# 1042|   		self_ptl = (ptl_t *) (ep->ptl_base_data + ptl_sizes);
# 1043|-> 		ptl_sizes += psmi_ptl_self.sizeof_ptl();
# 1044|   	}
# 1045|   

Error: CLANG_WARNING: [#def95]
libfabric-1.12.1/prov/psm3/psm3/psm_ep.c:1786:7: warning[core.NullDereference]: Array access (from variable 'devid_enabled') results in a null pointer dereference
# 1784|   	int i;
# 1785|   	for (i = 0; i < PTL_MAX_INIT; i++)
# 1786|-> 		if (devid_enabled[i] == devid)
# 1787|   			return 1;
# 1788|   	return 0;

Error: CLANG_WARNING: [#def96]
libfabric-1.12.1/prov/psm3/psm3/psm_ep_connect.c:248:7: warning[core.NullDereference]: Access to field 'ptlctl' results in a dereference of a null pointer
#  246|   
#  247|   		psmi_assert_always(array_of_epaddr[i] != NULL);
#  248|-> 		c = array_of_epaddr[i]->ptlctl;
#  249|   		psmi_assert_always(c != NULL);
#  250|   		_HFI_VDBG("%-20s DEVICE %s (%p)\n",

Error: TAINTED_SCALAR (CWE-20): [#def97]
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:338: tainted_argument: Calling function "hfi_sysfs_unit_read" taints argument "*cpulist".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:345: var_assign_var: Assigning: "temp" = "cpulist". Both are now tainted.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: tainted_data_transitive: Call to function "strchr" with tainted argument "*temp" returns tainted data.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: var_assign: Assigning: "next_comma" = "strchr(temp, 44)", which taints "next_comma".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:366: var_assign_var: Assigning: "temp" = "next_comma + 1". Both are now tainted.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: tainted_data_transitive: Call to function "strchr" with tainted argument "temp" returns tainted data.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: var_assign: Assigning: "next_comma" = "strchr(temp, 44)", which taints "next_comma".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:366: var_assign_var: Assigning: "temp" = "next_comma + 1". Both are now tainted.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:353: tainted_data_transitive: Call to function "atoi" with tainted argument "*temp" returns tainted data. [Note: The source code implementation of the function has been overridden by a builtin model.]
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:353: var_assign: Assigning: "first" = "atoi(temp)", which taints "first".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:361: var_assign_var: Assigning: "i" = "first". Both are now tainted.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:361: tainted_data: Using tainted variable "i" as a loop boundary.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:361: remediation: Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.
#  359|   			}
#  360|   
#  361|-> 			for (i = first; i <= last; i++) {
#  362|   				CPU_SET(i, cpuset);
#  363|   				ret++;

Error: TAINTED_SCALAR (CWE-20): [#def98]
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:338: tainted_argument: Calling function "hfi_sysfs_unit_read" taints argument "*cpulist".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:345: var_assign_var: Assigning: "temp" = "cpulist". Both are now tainted.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: tainted_data_transitive: Call to function "strchr" with tainted argument "*temp" returns tainted data.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: var_assign: Assigning: "next_comma" = "strchr(temp, 44)", which taints "next_comma".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:366: var_assign_var: Assigning: "temp" = "next_comma + 1". Both are now tainted.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: tainted_data_transitive: Call to function "strchr" with tainted argument "temp" returns tainted data.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:350: var_assign: Assigning: "next_comma" = "strchr(temp, 44)", which taints "next_comma".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:366: var_assign_var: Assigning: "temp" = "next_comma + 1". Both are now tainted.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:351: tainted_data_transitive: Call to function "strchr" with tainted argument "temp" returns tainted data.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:351: var_assign: Assigning: "dash" = "strchr(temp, 45)", which taints "dash".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:358: tainted_data_transitive: Call to function "atoi" with tainted argument "*dash" returns tainted data. [Note: The source code implementation of the function has been overridden by a builtin model.]
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:358: var_assign: Assigning: "last" = "atoi(dash + 1)", which taints "last".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:361: tainted_data: Using tainted variable "last" as a loop boundary.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:361: remediation: Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.
#  359|   			}
#  360|   
#  361|-> 			for (i = first; i <= last; i++) {
#  362|   				CPU_SET(i, cpuset);
#  363|   				ret++;

Error: RESOURCE_LEAK (CWE-772): [#def99] [important]
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:444: alloc_arg: "hfi_sysfs_port_read" allocates memory that is stored into "data_rate".
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:448: identity_transfer: Passing "data_rate" as argument 1 to function "strtod", which sets "newptr" to that argument.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:460: leaked_storage: Variable "newptr" going out of scope leaks the storage it points to.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/opa_service_gen1.c:460: leaked_storage: Variable "data_rate" going out of scope leaks the storage it points to.
#  458|   		  unit, port, strerror(errno));
#  459|   
#  460|-> 	return ret;
#  461|   }
#  462|   

Error: NO_EFFECT (CWE-398): [#def100]
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/psm_hal_inline_i.h:210: bad_memset: Calling "memset" with size 0: "memset(pc_private, 0, 0UL)" does nothing.
#  208|   	}
#  209|   
#  210|-> 	memset(pc_private, 0, sizeof(hfp_gen1_pc_private));
#  211|   
#  212|   

Error: CLANG_WARNING: [#def101]
libfabric-1.12.1/prov/psm3/psm3/psm.c:58: included_from: Included from here.
libfabric-1.12.1/prov/psm3/psm3/psm2_hal.h:365: included_from: Included from here.
libfabric-1.12.1/prov/psm3/psm3/psm2_hal_inlines_i.h:56: included_from: Included from here.
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/psm_hal_inline_i.h:214:7: warning[deadcode.DeadStores]: Although the value stored to 'err' is used in the enclosing expression, the value is never actually read from 'err'
#  212|   
#  213|   	// open verbs 1st so psmi_context_open can get pkey, lid, etc
#  214|-> 	if ((err = __psm2_ep_open_verbs(ep, unit, port, job_key)) != PSM2_OK) {
#  215|   		_HFI_ERROR( "Unable to initialize verbs\n");
#  216|   		err = -PSM_HAL_ERROR_CANNOT_OPEN_CONTEXT;

Error: CLANG_WARNING: [#def102]
libfabric-1.12.1/prov/psm3/psm3/psm_hal_gen1/psm_hal_inline_i.h:216:3: warning[deadcode.DeadStores]: Value stored to 'err' is never read
#  214|   	if ((err = __psm2_ep_open_verbs(ep, unit, port, job_key)) != PSM2_OK) {
#  215|   		_HFI_ERROR( "Unable to initialize verbs\n");
#  216|-> 		err = -PSM_HAL_ERROR_CANNOT_OPEN_CONTEXT;
#  217|   		goto bail;
#  218|   	}

Error: CLANG_WARNING: [#def103]
libfabric-1.12.1/prov/psm3/psm3/psm_mpool.c:453:2: warning[deadcode.DeadStores]: Value stored to 'nbytes' is never read
#  451|   	nbytes += mp->mp_elm_vector_size * sizeof(struct mpool_element *);
#  452|   	psmi_free(mp);
#  453|-> 	nbytes += sizeof(struct mpool);
#  454|   }
#  455|   

Error: ARRAY_VS_SINGLETON (CWE-119): [#def104]
libfabric-1.12.1/prov/psm3/psm3/psm_mq_recv.c:446: address_of: Taking address with "&offset" yields a singleton pointer.
libfabric-1.12.1/prov/psm3/psm3/psm_mq_recv.c:446: callee_ptr_arith: Passing "(uint32_t *)&offset" to function "mq_copy_tiny" which uses it as an array. This might corrupt or misinterpret adjacent memory locations.
#  444|   				 * copy after the DW payload.
#  445|   				 */
#  446|-> 				mq_copy_tiny((uint32_t *)((uint8_t *)user_buffer + paylen),
#  447|   					(uint32_t *)&offset, msglen & 0x3);
#  448|   			}

Error: ARRAY_VS_SINGLETON (CWE-119): [#def105]
libfabric-1.12.1/prov/psm3/psm3/psm_mq_recv.c:567: address_of: Taking address with "&offset" yields a singleton pointer.
libfabric-1.12.1/prov/psm3/psm3/psm_mq_recv.c:567: callee_ptr_arith: Passing "(uint32_t *)&offset" to function "mq_copy_tiny" which uses it as an array. This might corrupt or misinterpret adjacent memory locations.
#  565|   			 * copy after the DW payload.
#  566|   			 */
#  567|-> 			mq_copy_tiny((uint32_t *)(req->req_data.buf+paylen),
#  568|   				(uint32_t *)&offset, msglen & 0x3);
#  569|   		}

Error: CLANG_WARNING: [#def106]
libfabric-1.12.1/prov/psm3/psm3/psm_utils.c:56: included_from: Included from here.
libfabric-1.12.1/prov/psm3/psm3/psm_user.h:95: included_from: Included from here.
libfabric-1.12.1/prov/psm3/psm3/psm_utils.h:63: included_from: Included from here.
libfabric-1.12.1/prov/psm3/psm3/psm_netutils.h:82:31: warning[core.UndefinedBinaryOperatorResult]: The result of the left shift is undefined due to shifting by '32', which is greater or equal to the width of type 'unsigned int'
#   80|   static inline uint32_t psmi_bit_count_to_mask(int count)
#   81|   {
#   82|-> 	return (uint32_t)(0xffffffff << (32-count));
#   83|   }
#   84|   

Error: CLANG_WARNING: [#def107]
libfabric-1.12.1/prov/psm3/psm3/psm_timer.c:172:7: warning[core.NullDereference]: Access to field 't_timeout' results in a dereference of a null pointer
#  170|   	else
#  171|   		ctrl->t_cyc_next_expire =
#  172|-> 		    TAILQ_LAST(&ctrl->timerq, timerq)->t_timeout;
#  173|   
#  174|   	PSM2_LOG_MSG("leaving");

Error: CLANG_WARNING: [#def108]
libfabric-1.12.1/prov/psm3/psm3/psm_utils.c:335:7: warning[deadcode.DeadStores]: Although the value stored to 'h' is used in the enclosing expression, the value is never actually read from 'h'
#  333|   		return PSM2_OK;
#  334|   	/* First see if a hostname already exists */
#  335|-> 	if ((h = psmi_epid_lookup(PSMI_EP_HOSTNAME, nid)) != NULL) {
#  336|   		if (!overwrite)
#  337|   			return PSM2_OK;

Error: CLANG_WARNING: [#def109]
libfabric-1.12.1/prov/psm3/psm3/psm_utils.c:350:6: warning[unix.Malloc]: Potential leak of memory pointed to by 'h'
#  348|   	snprintf(h, hlen, "%s", hostname);
#  349|   	h[hlen - 1] = '\0';
#  350|-> 	err = psmi_epid_add(PSMI_EP_HOSTNAME, nid, h);
#  351|   	return err;
#  352|   }

Error: CLANG_WARNING: [#def110]
libfabric-1.12.1/prov/psm3/psm3/psm_utils.c:1422:26: warning[deadcode.DeadStores]: Value stored to 'entries' during its initialization is never read
# 1420|   void psmi_mem_stats_register(void)
# 1421|   {
# 1422|-> 	struct psmi_stats_entry entries[] = {
# 1423|   		PSMI_STATS_DECLU64("Total_(current)",
# 1424|   				(uint64_t*)&psmi_stats_memory.m_all_total),

Error: CLANG_WARNING: [#def111]
libfabric-1.12.1/prov/psm3/psm3/psm_utils.c:2123:10: warning[deadcode.DeadStores]: Although the value stored to 'err' is used in the enclosing expression, the value is never actually read from 'err'
# 2121|   	/* All AM options are read-only. */
# 2122|   	if (!get) {
# 2123|-> 		return err =
# 2124|   		    psmi_handle_error(PSMI_EP_LOGEVENT, PSM2_OPT_READONLY,
# 2125|   				      "Attempted to set read-only option value");

Error: CLANG_WARNING: [#def112]
libfabric-1.12.1/prov/psm3/psm3/psm_utils.c:2131:10: warning[deadcode.DeadStores]: Although the value stored to 'err' is used in the enclosing expression, the value is never actually read from 'err'
# 2129|   	if (*optlen < sizeof(uint32_t)) {
# 2130|   		*optlen = sizeof(uint32_t);
# 2131|-> 		return err = psmi_handle_error(PSMI_EP_LOGEVENT, PSM2_PARAM_ERR,
# 2132|   					       "Option value length error");
# 2133|   	}

Error: NEGATIVE_RETURNS (CWE-394): [#def113]
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_ep.c:420: negative_return_fn: Function "psmi_getpagesize()" returns a negative number.
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_ep.c:420: negative_returns: "psmi_getpagesize()" is passed to a parameter that cannot be negative.
#  418|   		pool->send_buffer_size = send_buffer_size;
#  419|   		pool->send_num_free = pool->send_total;
#  420|-> 		pool->send_buffers = (uint8_t *)psmi_memalign(ep, NETWORK_BUFFERS, CPU_PAGE_ALIGN,
#  421|   													 pool->send_total*pool->send_buffer_size);
#  422|   		if (! pool->send_buffers) {

Error: CLANG_WARNING: [#def114]
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_ep.c:644:42: warning[core.NullDereference]: Access to field 'allocator' results in a dereference of a null pointer (loaded from variable 'buf')
#  642|   			sbuf_t buf, uint32_t count)
#  643|   {
#  644|-> 	psm2_verbs_send_allocator_t allocator = buf->allocator;
#  645|   	psm2_verbs_send_pool_t pool = allocator->pool;
#  646|   	sbuf_t b;

Error: CLANG_WARNING: [#def115]
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_ep.c:791:10: warning[deadcode.DeadStores]: Value stored to 'buf' during its initialization is never read
#  789|   	// prepare RQ
#  790|   	for (i=0; i< pool->recv_total; i++) {
#  791|-> 		rbuf_t buf = &(pool->recv_bufs[i]);
#  792|   		if (pool->recv_buffer_size)
#  793|   			buf = &(pool->recv_bufs[i]);

Error: UNINIT (CWE-457): [#def116]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:217: var_decl: Declaring variable "act" without initializer.
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:220: uninit_use_in_call: Using uninitialized value "act.sa_mask" when calling "sigaction".
#  218|   				act.sa_sigaction = amsh_mmap_fault;
#  219|   				act.sa_flags = SA_SIGINFO;
#  220|-> 				psmi_sigaction(SIGSEGV, &act, NULL);
#  221|   			} else if (action_stash.SIGSEGV_old_act.sa_sigaction == (void*) SIG_IGN) {
#  222|   				return;

Error: UNINIT (CWE-457): [#def117]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:230: var_decl: Declaring variable "act" without initializer.
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:233: uninit_use_in_call: Using uninitialized value "act.sa_mask" when calling "sigaction".
#  231|   				act.sa_sigaction = amsh_mmap_fault;
#  232|   				act.sa_flags = SA_SIGINFO;
#  233|-> 				psmi_sigaction(SIGBUS, &act, NULL);
#  234|   			} else if (action_stash.SIGBUS_old_act.sa_sigaction == (void*) SIG_IGN) {
#  235|   				return;

Error: CONSTANT_EXPRESSION_RESULT (CWE-569): [#def118]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:272: result_independent_of_operands: "iterator <= 2147483647" is always true regardless of the values of its operands. This occurs as the logical operand of "for".
#  270|   
#  271|   	segsz = am_ctl_sizeof_block();
#  272|-> 	for (iterator = 0; iterator <= INT_MAX; iterator++) {
#  273|   		snprintf(shmbuf,
#  274|   			 sizeof(shmbuf),

Error: RESOURCE_LEAK (CWE-772): [#def119] [important]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:279: alloc_fn: Storage is returned from allocation function "psmi_strdup_internal".
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:279: var_assign: Assigning: "amsh_keyname" = storage returned from "psmi_strdup_internal(NULL, shmbuf, "prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:279")".
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:284: noescape: Resource "amsh_keyname" is not freed or pointed-to in "shm_open".
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:279: overwrite_var: Overwriting "amsh_keyname" in "amsh_keyname = psmi_strdup_internal(NULL, shmbuf, "prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:279")" leaks the storage that "amsh_keyname" points to.
#  277|   			 ep->epid,
#  278|   			 iterator);
#  279|-> 		amsh_keyname = psmi_strdup(NULL, shmbuf);
#  280|   		if (amsh_keyname == NULL) {
#  281|   			err = PSM2_NO_MEMORY;

Error: CLANG_WARNING: [#def120]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:429:2: warning[deadcode.DeadStores]: Value stored to 'shmidx' is never read
#  427|   	int iterator;
#  428|   
#  429|-> 	shmidx = *shmidx_o = -1;
#  430|   
#  431|   	for (i = 0; i <= ptl->max_ep_idx; i++) {

Error: CLANG_WARNING: [#def121]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:438:16: warning[deadcode.DeadStores]: Although the value stored to 'shmidx' is used in the enclosing expression, the value is never actually read from 'shmidx'
#  436|   				break;
#  437|   			}
#  438|-> 			*shmidx_o = shmidx = i;
#  439|   			return err;
#  440|   		}

Error: CONSTANT_EXPRESSION_RESULT (CWE-569): [#def122]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:447: result_independent_of_operands: "iterator <= 2147483647" is always true regardless of the values of its operands. This occurs as the logical operand of "for".
#  445|   
#  446|   	segsz = am_ctl_sizeof_block();
#  447|-> 	for (iterator = 0; iterator <= INT_MAX; iterator++) {
#  448|   		snprintf(shmbuf,
#  449|   			 sizeof(shmbuf),

Error: CLANG_WARNING: [#def123]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:559:6: warning[deadcode.DeadStores]: Value stored to 'use_kassist' is never read
#  557|   					ptl->psmi_kassist_mode =
#  558|   					    PSMI_KASSIST_OFF;
#  559|-> 					use_kassist = 0;
#  560|   					psmi_shm_mq_rv_thresh =
#  561|   					    PSMI_MQ_RV_THRESH_NO_KASSIST;

Error: CLANG_WARNING: [#def124]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:1058:6: warning[deadcode.DeadStores]: Value stored to 'err' is never read
# 1056|   				if (((am_epaddr_t *) epaddr)->cstate_incoming ==
# 1057|   					AMSH_CSTATE_INCOMING_DISC_REQUESTED)
# 1058|-> 					err = psmi_do_unmap(ptl->am_ep[shmidx].amsh_shmbase);
# 1059|   				req->epid_mask[i] = AMSH_CMASK_POSTREQ;
# 1060|   			} else if (req->epid_mask[i] == AMSH_CMASK_POSTREQ) {

Error: CLANG_WARNING: [#def125]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:1905:4: warning[deadcode.DeadStores]: Value stored to 'bulkptr' is never read
# 1903|   			break;
# 1904|   		default:
# 1905|-> 			bulkptr = 0;
# 1906|   			psmi_handle_error(PSMI_EP_NORETURN, PSM2_INTERNAL_ERR,
# 1907|   					  "Unknown/unhandled packet type 0x%x",

Error: CLANG_WARNING: [#def126]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2349:34: warning[core.NullDereference]: Access to field 'pid' results in a dereference of a null pointer (loaded from variable 'epaddr')
# 2347|   					     args_segoff);
# 2348|   
# 2349|-> 			((am_epaddr_t *) epaddr)->pid = pid;
# 2350|   			((am_epaddr_t *) epaddr)->gpuid = gpuid;
# 2351|   		}

Error: CLANG_WARNING: [#def127]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2423:5: warning[deadcode.DeadStores]: Value stored to 'err' is never read
# 2421|   			cstate = ((am_epaddr_t *) epaddr)->cstate_outgoing;
# 2422|   			if (cstate == AMSH_CSTATE_OUTGOING_DISC_REQUESTED) {
# 2423|-> 				err = psmi_do_unmap(ptl->am_ep[shmidx].amsh_shmbase);
# 2424|   				psmi_epid_remove(epaddr->ptlctl->ep, epaddr->epid);
# 2425|   			}

Error: UNUSED_VALUE (CWE-563): [#def128]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2658: value_overwrite: Overwriting previous write to "err" with value "PSM2_TIMEOUT".
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2670: value_overwrite: Overwriting previous write to "err" with value from "err_seg".
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2648: returned_value: Assigning value from "amsh_ep_disconnect(ptl_gen, force, num_disc, epaddr_array, mask, errs, timeout_ns)" to "err" here, but that stored value is overwritten before it can be used.
# 2646|   		psmi_epid_itor_fini(&itor);
# 2647|   		psmi_assert(i == num_disc && num_disc > 0);
# 2648|-> 		err = amsh_ep_disconnect(ptl_gen, force, num_disc, epaddr_array,
# 2649|   					 mask, errs, timeout_ns);
# 2650|   		psmi_free(mask);

Error: CLANG_WARNING: [#def129]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2648:3: warning[deadcode.DeadStores]: Value stored to 'err' is never read
# 2646|   		psmi_epid_itor_fini(&itor);
# 2647|   		psmi_assert(i == num_disc && num_disc > 0);
# 2648|-> 		err = amsh_ep_disconnect(ptl_gen, force, num_disc, epaddr_array,
# 2649|   					 mask, errs, timeout_ns);
# 2650|   		psmi_free(mask);

Error: UNUSED_VALUE (CWE-563): [#def130]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2670: value_overwrite: Overwriting previous write to "err" with value from "err_seg".
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2658: assigned_value: Assigning value "PSM2_TIMEOUT" to "err" here, but that stored value is overwritten before it can be used.
# 2656|   		while (ptl->connect_incoming > 0 || ptl->connect_outgoing > 0) {
# 2657|   			if (!psmi_cycles_left(t_start, timeout_ns)) {
# 2658|-> 				err = PSM2_TIMEOUT;
# 2659|   				_HFI_VDBG("CCC timed out with from=%d,to=%d\n",
# 2660|   					  ptl->connect_incoming, ptl->connect_outgoing);

Error: CLANG_WARNING: [#def131]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/am_reqrep_shmem.c:2658:5: warning[deadcode.DeadStores]: Value stored to 'err' is never read
# 2656|   		while (ptl->connect_incoming > 0 || ptl->connect_outgoing > 0) {
# 2657|   			if (!psmi_cycles_left(t_start, timeout_ns)) {
# 2658|-> 				err = PSM2_TIMEOUT;
# 2659|   				_HFI_VDBG("CCC timed out with from=%d,to=%d\n",
# 2660|   					  ptl->connect_incoming, ptl->connect_outgoing);

Error: CLANG_WARNING: [#def132]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/cmarwu.c:150:2: warning[deadcode.DeadStores]: Value stored to 'nr' is never read
#  148|   		.iov_len = n
#  149|   	};
#  150|-> 	nr = sum = 0;
#  151|   	while (sum != n) {
#  152|   		nr = process_vm_readv(pid, &local, 1, &remote, 1, 0);

Error: CLANG_WARNING: [#def133]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/cmarwu.c:177:2: warning[deadcode.DeadStores]: Value stored to 'nr' is never read
#  175|   	};
#  176|   
#  177|-> 	nr = sum = 0;
#  178|   	while (sum != n) {
#  179|   		nr = process_vm_writev(pid, &local, 1, &remote, 1, 0);

Error: CLANG_WARNING: [#def134]
libfabric-1.12.1/prov/psm3/psm3/ptl_am/ptl.c:216:3: warning[deadcode.DeadStores]: Value stored to 'rc' is never read
#  214|   	case MQ_MSG_SHORT:
#  215|   	case MQ_MSG_EAGER:
#  216|-> 		rc = psmi_mq_handle_envelope(tok->mq, tok->tok.epaddr_incoming,
#  217|   					     &tag, msglen, 0, buf,
#  218|   					     (uint32_t) len, 1, opcode, &req);

Error: UNINIT (CWE-457): [#def135]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:74: var_decl: Declaring variable "elid" without initializer.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:101: uninit_use_in_call: Using uninitialized value "elid". Field "elid.data" is uninitialized when calling "hsearch_r".
#   99|   		 query.slid, query.dlid);
#  100|   	elid.key = eplid;
#  101|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_rec_hash);
#  102|   
#  103|   	if (!epath) {		/* Unable to find path record in cache */

Error: CLANG_WARNING: [#def136]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:101:2: warning[core.CallAndMessage]: Passed-by-value struct argument contains uninitialized data (e.g., field: 'data')
#   99|   		 query.slid, query.dlid);
#  100|   	elid.key = eplid;
#  101|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_rec_hash);
#  102|   
#  103|   	if (!epath) {		/* Unable to find path record in cache */

Error: RESOURCE_LEAK (CWE-772): [#def137] [important]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:106: alloc_fn: Storage is returned from allocation function "psmi_calloc_internal".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:106: var_assign: Assigning: "path_rec" = storage returned from "psmi_calloc_internal(proto->ep, UNDEFINED, 1UL, 24UL, "prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:108")".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:206: leaked_storage: Variable "path_rec" going out of scope leaks the storage it points to.
#  204|   
#  205|   fail:
#  206|-> 	return err;
#  207|   }
#  208|   

Error: UNINIT (CWE-457): [#def138]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:221: var_decl: Declaring variable "elid" without initializer.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:292: uninit_use_in_call: Using uninitialized value "elid". Field "elid.data" is uninitialized when calling "hsearch_r".
#  290|   	snprintf(eplid, sizeof(eplid), "%x_%x", slid, dlid);
#  291|   	elid.key = eplid;
#  292|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_grp_hash);
#  293|   
#  294|   	if (epath) {		/* Find path group in cache */

Error: CLANG_WARNING: [#def139]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_opp_path_rec.c:292:2: warning[core.CallAndMessage]: Passed-by-value struct argument contains uninitialized data (e.g., field: 'data')
#  290|   	snprintf(eplid, sizeof(eplid), "%x_%x", slid, dlid);
#  291|   	elid.key = eplid;
#  292|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_grp_hash);
#  293|   
#  294|   	if (epath) {		/* Find path group in cache */

Error: UNINIT (CWE-457): [#def140]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:87: var_decl: Declaring variable "elid" without initializer.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:94: uninit_use_in_call: Using uninitialized value "elid". Field "elid.data" is uninitialized when calling "hsearch_r".
#   92|   	snprintf(eplid, sizeof(eplid), "%x_%x%04x", slid, ip_hi, dlid);
#   93|   	elid.key = eplid;
#   94|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_rec_hash);
#   95|   
#   96|   	if (!epath) {

Error: CLANG_WARNING: [#def141]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:94:2: warning[core.CallAndMessage]: Passed-by-value struct argument contains uninitialized data (e.g., field: 'data')
#   92|   	snprintf(eplid, sizeof(eplid), "%x_%x%04x", slid, ip_hi, dlid);
#   93|   	elid.key = eplid;
#   94|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_rec_hash);
#   95|   
#   96|   	if (!epath) {

Error: CLANG_WARNING: [#def142]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:259:17: warning[deadcode.DeadStores]: Value stored to 'num_path' during its initialization is never read
#  257|   {
#  258|   	psm2_error_t err = PSM2_OK;
#  259|-> 	uint16_t pidx, num_path = (1 << proto->epinfo.ep_lmc);
#  260|   	uint16_t path_slid, path_dlid;
#  261|   	ips_path_rec_t *path;

Error: UNINIT (CWE-457): [#def143]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:263: var_decl: Declaring variable "elid" without initializer.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:283: uninit_use_in_call: Using uninitialized value "elid". Field "elid.data" is uninitialized when calling "hsearch_r".
#  281|   	snprintf(eplid, sizeof(eplid), "%x_%x%04x", slid, ip_hi, dlid);
#  282|   	elid.key = eplid;
#  283|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_grp_hash);
#  284|   
#  285|   	if (epath) {		/* Find path group in cache */

Error: CLANG_WARNING: [#def144]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:283:2: warning[core.CallAndMessage]: Passed-by-value struct argument contains uninitialized data (e.g., field: 'data')
#  281|   	snprintf(eplid, sizeof(eplid), "%x_%x%04x", slid, ip_hi, dlid);
#  282|   	elid.key = eplid;
#  283|-> 	hsearch_r(elid, FIND, &epath, &proto->ips_path_grp_hash);
#  284|   
#  285|   	if (epath) {		/* Find path group in cache */

Error: DEADCODE (CWE-561): [#def145]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:292: assignment: Assigning: "num_path" = "1".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:317: const: At condition "num_path > 1", the value of "num_path" must be equal to 1.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:317: dead_error_condition: The condition "num_path > 1" cannot be true.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:319: dead_error_begin: Execution cannot reach this statement: "pathgrp->pg_num_paths[IPS_P...".
#  317|   	if (num_path > 1) {
#  318|   		/* One control path and (num_path - 1) norm and low priority paths */
#  319|-> 		pathgrp->pg_num_paths[IPS_PATH_HIGH_PRIORITY] = 1;
#  320|   		pathgrp->pg_num_paths[IPS_PATH_NORMAL_PRIORITY] = num_path - 1;
#  321|   		pathgrp->pg_num_paths[IPS_PATH_LOW_PRIORITY] = num_path - 1;

Error: DEADCODE (CWE-561): [#def146]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:317: cond_const: Condition "num_path > 1", taking false branch. Now the value of "num_path" is equal to 1.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:346: const: At condition "num_path > 1", the value of "num_path" must be equal to 1.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:346: dead_error_condition: The condition "num_path > 1" cannot be true.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_path_rec.c:347: dead_error_line: Execution cannot reach this statement: "if (pidx == 0) {
  pathgrp-...".
#  345|   
#  346|   		if (num_path > 1) {
#  347|-> 			if (pidx == 0) {
#  348|   				/* First path is always the high priority path */
#  349|   				pathgrp->pg_path[0][IPS_PATH_HIGH_PRIORITY] =

Error: CLANG_WARNING: [#def147]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto.c:853:3: warning[deadcode.DeadStores]: Value stored to 'err' is never read
#  851|   		}
#  852|   		psmi_epid_itor_fini(&itor);
#  853|-> 		err = ips_proto_disconnect(proto, force, num_disc, epaddr_array,
#  854|   					   mask, errs, timeout_in);
#  855|   		psmi_free(mask);

Error: CLANG_WARNING: [#def148]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto.c:1146:9: warning[core.CallAndMessage]: 8th function call argument is an uninitialized value
# 1144|   	case PSM_TRANSFER_PIO:
# 1145|   		GENERIC_PERF_BEGIN(PSM_TX_SPEEDPATH_CTR); /* perf stats */
# 1146|-> 		err = psmi_hal_spio_transfer_frame(proto, flow,
# 1147|   						   ctrlscb, payload, paylen,
# 1148|   						   PSMI_TRUE, have_cksum, ctrlscb->cksum[0],

Error: DEADCODE (CWE-561): [#def149]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:279: assignment: Assigning: "is_non_dw_mul_allowed" = "0U".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:290: cond_const: Condition "is_non_dw_mul_allowed", taking false branch. Now the value of "is_non_dw_mul_allowed" is equal to 0.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:290: const: At condition "is_non_dw_mul_allowed", the value of "is_non_dw_mul_allowed" must be equal to 0.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:290: dead_error_condition: The condition "is_non_dw_mul_allowed" cannot be true.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:294: dead_error_line: Execution cannot reach this statement: "padding = 0;".
#  292|   			 * multiple message size is allowed.
#  293|   			 */
#  294|-> 			padding = 0;
#  295|   		} else {
#  296|   			padding = nbytes_left & 0x3;

Error: NULL_RETURNS (CWE-476): [#def150]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:1477: returned_null: "psmi_mpool_get" returns "NULL" (checked 7 out of 8 times).
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:1477: var_assigned: Assigning: "sreq" = "NULL" return value from "psmi_mpool_get".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_mq.c:1481: dereference: Dereferencing "sreq", which is known to be "NULL".
libfabric-1.12.1/prov/psm3/psm3/psm_mq_utils.c:72: example_assign: Example 1: Assigning: "req" = return value from "psmi_mpool_get(mq->sreq_pool)".
libfabric-1.12.1/prov/psm3/psm3/psm_mq_utils.c:76: example_checked: Example 1 (cont.): "req" has its value checked in "req != NULL".
libfabric-1.12.1/prov/psm3/psm3/psm_mq_utils.c:74: example_assign: Example 2: Assigning: "req" = return value from "psmi_mpool_get(mq->rreq_pool)".
libfabric-1.12.1/prov/psm3/psm3/psm_mq_utils.c:76: example_checked: Example 2 (cont.): "req" has its value checked in "req != NULL".
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_mr.c:583: example_assign: Example 3: Assigning: "p_item" = return value from "psmi_mpool_get(cache->mr_pool)".
libfabric-1.12.1/prov/psm3/psm3/psm_verbs_mr.c:584: example_checked: Example 3 (cont.): "p_item" has its value checked in "p_item".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_am.c:566: example_assign: Example 4: Assigning: "msg" = return value from "psmi_mpool_get(ips_am_msg_pool)".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_am.c:567: example_checked: Example 4 (cont.): "msg" has its value checked in "msg == NULL".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_expected.c:434: example_assign: Example 5: Assigning: "getreq" = return value from "psmi_mpool_get(protoexp->tid_getreq_pool)".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_expected.c:439: example_checked: Example 5 (cont.): "getreq" has its value checked in "getreq == NULL".
# 1479|   			psmi_assert(sreq != NULL);
# 1480|   
# 1481|-> 			sreq->type = IPS_PENDSEND_EAGER_DATA;
# 1482|   			sreq->req = req;
# 1483|   			STAILQ_INSERT_TAIL(&proto->pend_sends.pendq, sreq, next);

Error: UNREACHABLE (CWE-561): [#def151]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ips_proto_recv.c:532: unreachable: This code cannot be reached: "ack_seq_num.psn_seq--;".
#  530|   		_HFI_ERROR("Got nak for TID flow, not allowed for UD\n");
#  531|   			goto ret;	/* Invalid ack for flow */
#  532|-> 		ack_seq_num.psn_seq--;
#  533|   
#  534|   		psmi_assert(flow->xmit_seq_num.psn_gen == ack_seq_num.psn_gen);

Error: IDENTICAL_BRANCHES (CWE-398): [#def152] [important]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ptl.c:273: implicit_else: The code from the above if-then branch is identical to the code after the if statement.
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ptl.c:271: identical_branches: The same code is executed when the condition "err = ips_ptl_rcvthread_init(ptl_gen, &ptl->recvq)" is true or false, because the code in the if-then branch and after the if statement is identical. Should the if statement be removed?
#  269|   	 * pthread.
#  270|   	 */
#  271|-> 	if ((err = ips_ptl_rcvthread_init(ptl_gen, &ptl->recvq)))
#  272|   		goto fail;
#  273|   fail:
```

**Comment by shefty** [comment]
Error: CLANG_WARNING: [#def153]
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ptl_rcvthread.c:451:23: warning[deadcode.DeadStores]: Value stored to 'recvq' during its initialization is never read
#  449|   {
#  450|   	struct ptl_rcvthread *rcvc = (struct ptl_rcvthread *)rcvthreadc;
#  451|-> 	struct ips_recvhdrq *recvq = rcvc->recvq;
#  452|   	int fd_pipe = rcvc->pipefd[0];
#  453|   	psm2_ep_t ep;

Error: DIVIDE_BY_ZERO (CWE-369): [#def154]
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:942: zero_return: Function call "ofi_datatype_size(datatype)" returns 0.
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:942: divide_by_zero: In expression "len / ofi_datatype_size(datatype)", division by expression "ofi_datatype_size(datatype)" which may be zero has undefined behavior.
#  940|   		psmx3_ioc_read(iov, count, datatype, buf, len);
#  941|   
#  942|-> 		err = psmx3_atomic_self(PSMX3_AM_REQ_ATOMIC_WRITE, ep_priv,
#  943|   					buf, len / ofi_datatype_size(datatype),
#  944|   					NULL, NULL, NULL, NULL, NULL, addr,

Error: DIVIDE_BY_ZERO (CWE-369): [#def155]
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:986: zero_return: Function call "ofi_datatype_size(datatype)" returns 0.
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:986: divide_by_zero: In expression "len / ofi_datatype_size(datatype)", division by expression "ofi_datatype_size(datatype)" which may be zero has undefined behavior.
#  984|   
#  985|   	args[0].u32w0 = PSMX3_AM_REQ_ATOMIC_WRITE;
#  986|-> 	args[0].u32w1 = len / ofi_datatype_size(datatype);
#  987|   	args[1].u64 = (uint64_t)(uintptr_t)req;
#  988|   	args[2].u64 = addr;

Error: FORWARD_NULL (CWE-476): [#def156]
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1241: var_compare_op: Comparing "iov" to null implies that "iov" might be null.
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1252: var_deref_op: Dereferencing null pointer "iov".
# 1250|   
# 1251|   	if (op != FI_ATOMIC_READ) {
# 1252|-> 		buf = iov[0].addr; /* as default for count == 1 */
# 1253|   		len = psmx3_ioc_size(iov, count, datatype);
# 1254|   		desc0 = desc ? desc[0] : NULL;

Error: DIVIDE_BY_ZERO (CWE-369): [#def157]
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1289: zero_return: Function call "ofi_datatype_size(datatype)" returns 0.
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1289: divide_by_zero: In expression "len / ofi_datatype_size(datatype)", division by expression "ofi_datatype_size(datatype)" which may be zero has undefined behavior.
# 1287|   		}
# 1288|   
# 1289|-> 		err = psmx3_atomic_self(PSMX3_AM_REQ_ATOMIC_READWRITE, ep_priv,
# 1290|   					buf, len / ofi_datatype_size(datatype),
# 1291|   					desc0, NULL, NULL, result, result_desc0,

Error: DIVIDE_BY_ZERO (CWE-369): [#def158]
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1359: zero_return: Function call "ofi_datatype_size(datatype)" returns 0.
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1359: divide_by_zero: In expression "len / ofi_datatype_size(datatype)", division by expression "ofi_datatype_size(datatype)" which may be zero has undefined behavior.
# 1357|   
# 1358|   	args[0].u32w0 = PSMX3_AM_REQ_ATOMIC_READWRITE;
# 1359|-> 	args[0].u32w1 = len / ofi_datatype_size(datatype);
# 1360|   	args[1].u64 = (uint64_t)(uintptr_t)req;
# 1361|   	args[2].u64 = addr;

Error: DIVIDE_BY_ZERO (CWE-369): [#def159]
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1706: zero_return: Function call "ofi_datatype_size(datatype)" returns 0.
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1706: divide_by_zero: In expression "len / ofi_datatype_size(datatype)", division by expression "ofi_datatype_size(datatype)" which may be zero has undefined behavior.
# 1704|   		}
# 1705|   
# 1706|-> 		err = psmx3_atomic_self(PSMX3_AM_REQ_ATOMIC_COMPWRITE, ep_priv,
# 1707|   					buf, len / ofi_datatype_size(datatype), desc0,
# 1708|   					compare, compare_desc0, result, result_desc0,

Error: DIVIDE_BY_ZERO (CWE-369): [#def160]
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1777: zero_return: Function call "ofi_datatype_size(datatype)" returns 0.
libfabric-1.12.1/prov/psm3/src/psmx3_atomic.c:1777: divide_by_zero: In expression "len / ofi_datatype_size(datatype)", division by expression "ofi_datatype_size(datatype)" which may be zero has undefined behavior.
# 1775|   
# 1776|   	args[0].u32w0 = PSMX3_AM_REQ_ATOMIC_COMPWRITE;
# 1777|-> 	args[0].u32w1 = len / ofi_datatype_size(datatype);
# 1778|   	args[1].u64 = (uint64_t)(uintptr_t)req;
# 1779|   	args[2].u64 = addr;

Error: DEADCODE (CWE-561): [#def161]
libfabric-1.12.1/prov/psm3/src/psmx3_av.c:448: assignment: Assigning: "error_count" = "0".
libfabric-1.12.1/prov/psm3/src/psmx3_av.c:515: const: At condition "error_count", the value of "error_count" must be equal to 0.
libfabric-1.12.1/prov/psm3/src/psmx3_av.c:515: dead_error_condition: The condition "error_count" cannot be true.
libfabric-1.12.1/prov/psm3/src/psmx3_av.c:516: dead_error_begin: Execution cannot reach this statement: "for (i = 0; i < count; i++)...".
libfabric-1.12.1/prov/psm3/src/psmx3_av.c:516: effectively_constant: Local variable "error_count" is assigned only once, to a constant value, making it effectively constant throughout its scope. If this is not the intent, examine the logic to see if there is a missing assignment that would make "error_count" not remain constant.
#  514|   	if (av_priv->flags & FI_EVENT) {
#  515|   		if (error_count) {
#  516|-> 			for (i = 0; i < count; i++)
#  517|   				psmx3_av_post_completion(av_priv, context, i, errors[i]);
#  518|   		}

Error: FORWARD_NULL (CWE-476): [#def162]
libfabric-1.12.1/prov/psm3/src/psmx3_cq.c:851: var_compare_op: Comparing "buf" to null implies that "buf" might be null.
libfabric-1.12.1/prov/psm3/src/psmx3_cq.c:915: var_deref_model: Passing null pointer "buf" to "memcpy", which dereferences it. [Note: The source code implementation of the function has been overridden by a builtin model.]
#  913|   				}
#  914|   
#  915|-> 				memcpy(buf, (void *)&event->cqe, cq_priv->entry_size);
#  916|   				psmx3_cq_free_event(cq_priv, event);
#  917|   

Error: CLANG_WARNING: [#def163]
libfabric-1.12.1/prov/psm3/src/psmx3_cq.c:915:5: warning[core.NonNullParamChecker]: Null pointer passed to 1st parameter expecting 'nonnull'
#  913|   				}
#  914|   
#  915|-> 				memcpy(buf, (void *)&event->cqe, cq_priv->entry_size);
#  916|   				psmx3_cq_free_event(cq_priv, event);
#  917|   

Error: CHECKED_RETURN (CWE-252): [#def164]
libfabric-1.12.1/prov/psm3/src/psmx3_domain.c:492: check_return: Calling "ofi_domain_close" without checking return value (as is done elsewhere 14 out of 17 times).
libfabric-1.12.1/prov/efa/src/efa_domain.c:76: example_assign: Example 1: Assigning: "ret" = return value from "ofi_domain_close(&domain->util_domain)".
libfabric-1.12.1/prov/efa/src/efa_domain.c:77: example_checked: Example 1 (cont.): "ret" has its value checked in "ret".
libfabric-1.12.1/prov/efa/src/rxr/rxr_domain.c:71: example_assign: Example 2: Assigning: "ret" = return value from "ofi_domain_close(&rxr_domain->util_domain)".
libfabric-1.12.1/prov/efa/src/rxr/rxr_domain.c:72: example_checked: Example 2 (cont.): "ret" has its value checked in "ret".
libfabric-1.12.1/prov/mrail/src/mrail_domain.c:49: example_assign: Example 3: Assigning: "ret" = return value from "ofi_domain_close(&mrail_domain->util_domain)".
libfabric-1.12.1/prov/mrail/src/mrail_domain.c:50: example_checked: Example 3 (cont.): "ret" has its value checked in "ret".
libfabric-1.12.1/prov/psm2/src/psmx2_domain.c:187: example_checked: Example 4: "ofi_domain_close(&domain->util_domain)" has its value checked in "ofi_domain_close(&domain->util_domain)".
libfabric-1.12.1/prov/psm3/src/psmx3_domain.c:187: example_checked: Example 5: "ofi_domain_close(&domain->util_domain)" has its value checked in "ofi_domain_close(&domain->util_domain)".
#  490|   
#  491|   err_out_close_domain:
#  492|-> 	ofi_domain_close(&domain_priv->util_domain);
#  493|   
#  494|   err_out_free_domain:

Error: CLANG_WARNING: [#def165]
libfabric-1.12.1/prov/psm3/src/psmx3_msg.c:259:31: warning[core.NullDereference]: Array access (via field 'i') results in a null pointer dereference
#  257|   		assert(context);
#  258|   		fi_context = context;
#  259|-> 		PSMX3_CTXT_TYPE(fi_context) = PSMX3_SEND_CONTEXT;
#  260|   		PSMX3_CTXT_USER(fi_context) = (void *)buf;
#  261|   		PSMX3_CTXT_EP(fi_context) = ep_priv;

Error: CLANG_WARNING: [#def166]
libfabric-1.12.1/prov/psm3/src/psmx3_tagged.c:584:31: warning[core.NullDereference]: Array access (via field 'i') results in a null pointer dereference
#  582|   		assert(context);
#  583|   		fi_context = context;
#  584|-> 		PSMX3_CTXT_TYPE(fi_context) = PSMX3_TSEND_CONTEXT;
#  585|   		PSMX3_CTXT_USER(fi_context) = (void *)buf;
#  586|   		PSMX3_CTXT_EP(fi_context) = ep_priv;

Error: CLANG_WARNING: [#def167]
libfabric-1.12.1/prov/psm3/src/psmx3_tagged.c:34: included_from: Included from here.
libfabric-1.12.1/prov/psm3/src/psmx3_trigger.h:598:31: warning[core.NullDereference]: Dereference of null pointer
#  596|   
#  597|   	trigger->op = PSMX3_TRIGGERED_TSEND;
#  598|-> 	trigger->cntr = container_of(ctxt->trigger.threshold.cntr,
#  599|   				     struct psmx3_fid_cntr, cntr);
#  600|   	trigger->threshold = ctxt->trigger.threshold.threshold;

Error: CLANG_WARNING: [#def168]
libfabric-1.12.1/prov/psm3/src/psmx3_msg.c:34: included_from: Included from here.
libfabric-1.12.1/prov/psm3/src/psmx3_trigger.h:688:31: warning[core.NullDereference]: Dereference of null pointer
#  686|   
#  687|   	trigger->op = PSMX3_TRIGGERED_SEND;
#  688|-> 	trigger->cntr = container_of(ctxt->trigger.threshold.cntr,
#  689|   				     struct psmx3_fid_cntr, cntr);
#  690|   	trigger->threshold = ctxt->trigger.threshold.threshold;

Error: CHECKED_RETURN (CWE-252): [#def169]
libfabric-1.12.1/prov/psm3/src/psmx3_trx_ctxt.c:111: check_return: Calling "pthread_create" without checking return value (as is done elsewhere 15 out of 18 times).
libfabric-1.12.1/prov/psm2/src/psmx2_domain.c:141: example_assign: Example 1: Assigning: "err" = return value from "pthread_create(&domain->progress_thread, NULL, psmx2_progress_func, (void *)domain)".
libfabric-1.12.1/prov/psm2/src/psmx2_domain.c:143: example_checked: Example 1 (cont.): "err" has its value checked in "err".
libfabric-1.12.1/prov/psm2/src/psmx2_wait.c:118: example_assign: Example 2: Assigning: "err" = return value from "pthread_create(&psmx2_wait_thread, &attr, psmx2_wait_progress, (void *)fabric)".
libfabric-1.12.1/prov/psm2/src/psmx2_wait.c:120: example_checked: Example 2 (cont.): "err" has its value checked in "err".
libfabric-1.12.1/prov/psm3/psm3/psm_mq.c:1413: example_checked: Example 3: "pthread_create(&mq->mq_perf_data.perf_print_thread, NULL, psmi_mq_print_stats_thread, (void *)mq)" has its value checked in "pthread_create(&mq->mq_perf_data.perf_print_thread, NULL, psmi_mq_print_stats_thread, (void *)mq)".
libfabric-1.12.1/prov/psm3/psm3/psm_stats.c:289: example_checked: Example 4: "pthread_create(&perf_print_thread, NULL, psmi_print_stats_thread, NULL)" has its value checked in "pthread_create(&perf_print_thread, NULL, psmi_print_stats_thread, NULL)".
libfabric-1.12.1/prov/psm3/psm3/ptl_ips/ptl_rcvthread.c:145: example_checked: Example 5: "pthread_create(&rcvc->hdrq_threadid, NULL, ips_ptl_pollintr, ptl->rcvthread)" has its value checked in "pthread_create(&rcvc->hdrq_threadid, NULL, ips_ptl_pollintr, ptl->rcvthread)".
#  109|   			disconn->trx_ctxt = trx_ctxt;
#  110|   			disconn->epaddr = epaddr;
#  111|-> 			pthread_create(&disconnect_thread, NULL,
#  112|   				       disconnect_func, disconn);
#  113|   			pthread_detach(disconnect_thread);

**Comment by acgoldma** [comment]
This should be fixed, closing.
